### PR TITLE
Validation aggregation order

### LINKED ISSUE #67: Order of validation & aggregation
This issue is in response to openENTRANCE/openentrance#130.

There is an issue that can occur related to the order of operations for validation and region aggregation.
The way these two steps were originally supposed to be run is like this:

``` python
dsd.validate(df)
df = rp.apply(df)
```

That brings a problem because `df` contains might contain regions that are not in the DataStructureDefinition. That might not be a problem though as these regions might be dropped or renamed as part of the region processing. 
This brought the following workaround:

``` python
# only validate the variables (or any other non region dimensions) at this point
dsd.validate(df, dimensions=["variable"]) 
df = rp.apply(df)
```

This works as the model mapping is checked in the region processor, upon instantiation so any illegal regions are going to be detected. There is, however, one case in which this falls short. If we have a model for which we don't have a model mapping defined nothing is checked so we would still need to verify the resulting data frame explicitly.

``` python
# only validate the variables (or any other non region dimensions) at this point
dsd.validate(df, dimensions=["variable"]) 
df = rp.apply(df)
dsd.validate(df, dimensions=["region"]) 
```

This is not the cleanest implementation and can easily lead to wrong errors or illegal regions in the data base. 

The crux of the problem is that the question whether or not the `DataStructureDefinition` should validate the regions of the incoming `df` directly hinges on whether or not a model mapping is defined for the data inside `df`:

* If there **is** a mapping, the `RegionProcessor` has already taken care of this step so the incoming regions **must not** be validated. 
* If there is **no** mapping, the regions **must** be validated by `DataStructureDefinition` as there is no other validation step.

The information of the existence of a model mapping is **only** available to the `RegionProcessor`, **not** the `DataStructureDefinition`. 

Therefore I would propose two different solutions:

1. Move the validation functionality also inside `RegionProcessor`. By creating `RegionProcessor.validate()` we can iterate through the models inside the dataframe and decide on a per model basis whether or not we validate the regions. This would give something like this:
  ```python
  rp.validate(df)
  df = rp.apply(df)
  ```
2. Alternatively, we can make region validation part of `RegionProcessor.apply()`. This would take minimal effort as there is already an if clause in `RegionProcessor.apply()` that checks whether or not a model mapping is present. As of right now it only logs the fact that there is none and then returns the data unchanged. Adding region validation here would be a single line. With this the validation process would look like this: 
```python
dsd.validate(df) # per default anything except regions
df = rp.apply(df) # apply region check if no model mapping is found
```

What are your thoughts @danielhuppmann? I hope I got the description of the problem correct.

### PR DESCRIPTION
**Pull request by phackstock**
closes #67.

Changes:

* Created a new `core.py` which holds the `process` function.
* Moved tests that affect `process` from `test_region_aggregation` to `test_core`. Side benefit, now `test_region_aggregation` does not need to import `DataStructureDefinition` anymore, suggesting cleaner separation of the two classes.
* Found a unit test `test_region_processing_complete` that should have failed in the past but wrongly did not. Using the new `process` function this error was found and fixed. 
* Changed `testing.py` accordingly to reflect the new way to instantiate `RegionProcessor` (without `DataStructureDefinition`).
* Added functions to validate the mappings in `RegionAggregationMapping` (for a single mapping) and in `RegionProcessor` to validate all mappings against a `DataStructureDefinition`. *Side note*: The function name `validate` is part of the `BaseModel` parent class that both `RegionProcessor` and `DataStructureDefinition` inherit from so overwriting it would have broken the class. I went for `RegionProcessor.validate_mappings` and `DataStructureDefinition.validate_against_dsd`. Names can for sure be changed. 
* I left validating model regions where no mapping is defined out of `RegionProcessor.apply` as discussed in #67. This means that in `core:process` an additional `dsd.validate(df, dimensions=["region"])` is required. 
* In `RegionProcessor.apply` the mapping is only validated for the current model and not for all models in the project. This avoids the case where we would try to aggregate a data frame containing only 'model_a' and we get an error because a mapping for 'model_b' is broken. This is done by using the new `RegionAggregationMapping.validate_against_dsd` which only validates a single `RegionAggregationMapping` instead of using `RegionProcessor.validate_mappings` which does all.
* There is still a safeguard in place though against faulty mappings in general. `RegionProcessor.validate_mappings` is still called as part of `testing:assert_valid_structure`. 

Optional additions:

* There are currently only working test cases for `core:process`. Maybe we should add some failing ones as well.

Looking forward to your comments @danielhuppmann

### CODE REVIEW

**Comment by phackstock** [comment]
Two more comments from my side:
* As I moved the validation of the mappings for `RegionProcessor` into a function which is not a normal validation function from pydantic (i.e. invoked at creation), I had to find a way to collect errors and raise them all at once. I was able to use the pydantic `ValidationError` like this:
```python
errors = []
for mapping in self.mappings.values():
    try:
        mapping.validate_against_dsd(dsd)
    except RegionNotDefinedError as rnde:
        errors.append(ErrorWrapper(rnde, f"mappings -> {mapping.model}"))
    if errors:
        raise pydantic.ValidationError(errors, model=self.__class__)
```
This way all the errors are collected and then raised as a single `pydantic.ValidationError`. This could also be used in `testing:assert_valid_yaml` and `validation:validate`, instead of having to log the errors and refer the user to the logs. 
* When I implemented the `process` function in `core.py` I noticed an interesting bug:
```python
def process(
    df: IamDataFrame,
    dsd: DataStructureDefinition,
    dimensions: list = None,
    processor: RegionProcessor = None,
):
    dimensions = dimensions or dsd.dimensions
    if processor is None:
        return dsd.validate(df, dimensions=dimensions)
    if "region" in dimensions:
        dimensions.remove("region")
        dsd.validate(df, dimensions=dimensions)
        df = processor.apply(df, dsd)
        dsd.validate(df, dimensions=["region"])
    else:
        dsd.validate(df, dimensions=dimensions)
        df = processor.apply(df, dsd)
    return df
```
 If I did not specify a `dimensions` parameter, the `dimensions` variable would be `dsd.dimensions`. Since`dimensions.remove()` modifies a list in place rather than returning a copy, this **permanently** removes it from the dsd object. This change persists also outside of the scope of the function. There are two solutions that I see. 1. run `dsd.dimensions.append("region")` after the validation to add the region back in or 2. create a deep copy at the beginning of the function so that `dsd` is not actually modified `dimensions = copy.deepcopy(dimensions or dsd.dimensions)`. I ended up going for option 2.
Just though it was worth mentioning it as a bit of a cautionary tale since that bug caused weird behavior in testing. When I would run all the tests some would fail. If I ran the same tests on their own, they would run just fine. My hypothesis is that the reason for this was a combination of the fact that the `simple_definition` fixture has a session scope (and not function) and was modified in place by the units tests if run all in succession.

**Review by danielhuppmann — COMMENTED** [review]
Thanks @phackstock, very nice work!

I think it's fine to have only working tests for `process()` because all individual components used in the function are also extensively tested.

One question for clarification: currently, the method `validate_against_dsd()` (which I suggest to rename to `validation_regions()`, see inline comment) passes if the **DataStructureDefinition** does not have a "region" dimension. I assume that you did think about this and it's a conscious decision? (Not against the decision, but just spent a few minutes thinking about this and wanted to double-check...)

File: nomenclature/processor/region.py [hunk 1]
```diff
@@ -239,28 +241,27 @@ def model_native_region_names(self) -> List[str]:
     def rename_mapping(self) -> Dict[str, str]:
         return {r.name: r.target_native_region for r in self.native_regions or []}
 
+    def validate_against_dsd(self, dsd: DataStructureDefinition) -> None:
```
[the remark this answers is not in the document]
Comment (phackstock) [reply nomenclature/processor/region.py:244]: Ah yes, good suggestion, I wasn't really happy with the name in the first place.

**Comment by phackstock** [comment]
Good question about the `validate_regions` that passes if `DataStructureDefinition` does no have a region dimension. 
It was a conscious decision, just not sure if it was a good one.
Here's my reasoning:
Since we run `validate_regions` as part of `RegionProcessor.apply()` it would fail if the `DataStructureDefinition` does not have a region dimension. I though there might be a case where we use a `DataStructureDefinition` that does not have a region dimension and we still want to preform aggregation.
It might be a bit dangerous though. Alternatively I could add a `validate_region` parameter to `RegionProcessor.apply()` which defaults to true as well as change `validate_regions` to fail if `DataStructureDefinition` does not have regions.
Might be the safer option.

**Comment by danielhuppmann** [comment]
Ok, we're on the same page - it's fine to not validate for regions if this dimension doesn't exist in the DataStructureDefinition.

This supports a minimal operation where the region-processor can be used with a variable-definitions (to know how to apply processing, which methods, etc.) and a region-mapping, without the need to duplicate the list of regions as a definition.

**Comment by phackstock** [comment]
Ok, so we're good to merge then without any more changes?

**Review by danielhuppmann — APPROVED** [review]
Thanks @phackstock, looks great!

**Comment by phackstock** [comment]
I guess depending on whether all the projects that use nomenclature use it from pip or github, we could adjust those projects to use the process function next.

**Pull request merged**